cmd/zoekt-sourcegraph-indexserver/sg.go

```diff
@@ -0,0 +1,162 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	retryablehttp "github.com/hashicorp/go-retryablehttp"
+)
+
+// Sourcegraph contains methods which interact with the Sourcegraph API.
+type Sourcegraph struct {
+	// Root is the base URL for the Sourcegraph instance to index. Normally
+	// http://sourcegraph-frontend-internal or http://localhost:3090.
+	Root *url.URL
+
+	// Hostname is the name we advertise to Sourcegraph when asking for the
+	// list of repositories to index.
+	Hostname string
+
+	Client *retryablehttp.Client
+}
+
+// indexOptionsItem wraps IndexOptions to also include an error returned by
+// the API.
+type indexOptionsItem struct {
+	IndexOptions
+	Error string
+}
+
+func (s *Sourcegraph) GetIndexOptions(repos ...string) ([]indexOptionsItem, error) {
+	u := s.Root.ResolveReference(&url.URL{
+		Path: "/.internal/search/configuration",
+	})
+
+	resp, err := s.Client.PostForm(u.String(), url.Values{"repo": repos})
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		b, err := ioutil.ReadAll(io.LimitReader(resp.Body, 1024))
+		_ = resp.Body.Close()
+		if err != nil {
+			return nil, err
+		}
+		return nil, &url.Error{
+			Op:  "Get",
+			URL: u.String(),
+			Err: fmt.Errorf("%s: %s", resp.Status, string(b)),
+		}
+	}
+
+	opts := make([]indexOptionsItem, len(repos))
+	dec := json.NewDecoder(resp.Body)
+	for i := range opts {
+		if err := dec.Decode(&opts[i]); err != nil {
+			return nil, fmt.Errorf("error decoding body: %w", err)
+		}
+	}
+
+	return opts, nil
+}
+
+func (s *Sourcegraph) GetCloneURL(name string) string {
+	return s.Root.ResolveReference(&url.URL{Path: path.Join("/.internal/git", name)}).String()
+}
+
+func (s *Sourcegraph) WaitForFrontend() {
```

Comment:
Since we're using an HTTP client with retries, I think we can avoid waiting for frontends. It's also a bit of an anti-pattern I think, since availability isn't really constant, so retrying individual requests is a better approach I think.

I got rid of our WaitForX usages in Sourcegraph recently: https://github.com/sourcegraph/sourcegraph/pull/23575

Reply:
Context: If I remember the history, we did this to avoid log spam on startup. IE the retry policy on startup needs to wait much longer than we want normal operations to run. This was specifically targetting `server` docker image which may run migrations/wait for postgres to start/etc before frontend was avail. I wonder if we have removed the WaitForX in other places we might have added a lot of logspam in server?

I'm happy to remove this in another PR.

Reply:
>  I wonder if we have removed the WaitForX in other places we might have added a lot of logspam in server?

We don't because the retries in the HTTP client don't log until they fail, and the default retry attempt count is 20, with exponential back-off.

Reply:
https://github.com/sourcegraph/sourcegraph/blob/afa68a400bd8e2affa83095024c07d3bcefc6f3f/internal/httpcli/client.go#L127

Reply:
Cool, then I'll remove this call.